src/LanguageServer/Impl/Implementation/LineFormatter.cs

```diff
@@ -0,0 +1,631 @@
+        /// </summary>
+        /// <param name="reader">The code to be formatted. LineFormatter does not dispose of the reader.</param>
+        /// <param name="languageVersion">Language version to use in the tokenization format.</param>
+        public LineFormatter(TextReader reader, PythonLanguageVersion languageVersion) {
+            var tokenizer = new Tokenizer(languageVersion, options: TokenizerOptions.Verbatim | TokenizerOptions.VerbatimCommentsAndLineJoins | TokenizerOptions.GroupingRecovery);
+            tokenizer.Initialize(reader);
+            _tokenizer = new TokenizerWrapper(tokenizer);
+            _lineTokens = new Dictionary<int, List<TokenExt>>();
+        }
+
+        private void AddToken(TokenExt token) {
+            var line = token.Line;
+
+            // Explicit line joins ("\") appear at the end of a line, but
+            // their span ends on another line, so move backward so they can
+            // be inserted in the right place.
+            if (token.Kind == TokenKind.ExplicitLineJoin) {
+                line--;
+            }
+
+            if (!_lineTokens.TryGetValue(line, out List<TokenExt> tokens)) {
+                tokens = new List<TokenExt>();
+                _lineTokens.Add(line, tokens);
+            }
+
+            tokens.Add(token);
+        }
+
+        /// <summary>
+        /// Tokenizes up to and including the specified line. Tokens are
+        /// stored in _lineTokens. If the provided line number is past the
+        /// end of the input text, then the tokenizer will stop.
+        /// Additionally, this function will attempt to read ahead onto the
+        /// next line to the first non-ignored token so that the formatter
+        /// can look ahead.
+        /// </summary>
+        /// <param name="line">One-indexed line number.</param>
+        /// <returns>A non-null list of tokens on that line.</returns>
+        private List<TokenExt> TokenizeLine(int line) {
+            Check.Argument(nameof(line), () => line > 0);
+
+            var extraToken = true;
+
+            var peeked = _tokenizer.Peek();
+            while (peeked != null && (peeked.Line <= line || extraToken)) {
+                var token = _tokenizer.Next();
+                AddToken(token);
+                peeked = _tokenizer.Peek();
+
+                if (token.Line > line && !token.IsIgnored) {
+                    extraToken = false;
+                }
+            }
+
+            if (!_lineTokens.TryGetValue(line, out List<TokenExt> tokens)) {
+                return new List<TokenExt>();
+            }
+
+            return tokens;
+        }
+
+        /// <summary>
+        /// Formats a single line and returns TextEdits to replace the old text.
+        /// </summary>
+        /// <param name="line">One-indexed line number.</param>
+        /// <returns>A list of TextEdits needed to format the line.</returns>
+        public TextEdit[] FormatLine(int line) {
```

Comment:
This method is the only public one. Is it assumed that it can be called only once?